### DISCUSSION TITLE
Panel to display the compass accuracy

### DISCUSSION THREAD

**Discussion by kds112** [devemux86/cruiser#667] [Q&A] [answered]
As discussed in #625 Cruiser 5.0.0 and 5.0.1 made a big step for hikers using the internal compass for orientation on the map. I just want to put one last feature on Emux' implemetation list... An optional panel that indicates the current compass accuracy, maybe with three bars for high, medium, low. In the low state the color might change to red similar to the battery panel.
Thanks a lot! 🧭

**Comment by devemux86** [comment]
@kds112 Thanks for the suggestion.

Panels display text to inform about their value. This way they are also much more readable.
And the text can be customized in terms of size and color.

**Reply by kds112** [reply]
Hi/med/lo as text is as good as bars... 🙂

**Reply by devemux86** [reply]
We have to think all the supported languages.
Something small in English doesn't work in other languages.

Numbers or percentage are better in such cases.

**Reply by kds112** [reply]
Similar to the panel for gps accuracy (numbers of satellites) you could use "3/3" (high), "2/3" (medium) and "1/3" (low)...?

**Reply by kds112** [reply]
Sorry, I was too slow 😉

**Reply by devemux86** [reply]
Same thought! :slightly_smiling_face:

**Comment by Ciutnik** [comment]
I don't question the need for such a panel. 

But it's worth getting used to calibrating the compass every time you use an app that uses it. 
It takes two seconds to make one or two "figures eight" in the air with your phone. 
I know this from my own experience, most recently three days ago 🙂

Of course I mean the "walker mode" 🙂

**Comment by devemux86** [comment]
New panel to display the compass accuracy will be available in the next Cruiser.

Compass accuracy steps are fixed (0/3, 1/3, 2/3, 3/3) by Android.

<img src="https://github.com/user-attachments/assets/292d91e2-9340-4827-a289-be1f61ecb60a" width="50%"/>

**Reply by kds112** [reply]
Sounds really good 👍 Will 66% be in yellow/orange and 33%/0% in red text? Similar to the battery panel?

**Reply by devemux86** [reply]
Battery is a special case and only has red as extra.

Fixed text colors can be confusing, as there is the option to select a custom text color for all panels.

**Reply by kds112** [reply]
Ok, I see... 👍🧭

**Reply by bios4** [reply]
> Accuracy steps are fixed (0/3, 1/3, 2/3, 3/3) by Android, we could display them as percentage (0%, 33%, 66%, 100%)?

Why not simply displaying them exactly like Android?

**Reply by devemux86** [reply]
We are discussing how to display them (Android does not display them).
Satellites 7/15 have a understandable meaning.

Is compass accuracy 2/3 understandable or 66% is better?
How GPS status apps display the compass accuracy value?

**Reply by kds112** [reply]
I think 2/3 is better than 66%... My compass app uses 3 bars and the magnetic field strength in microTesla:
![IMG_20241111_202759](https://github.com/user-attachments/assets/2088f796-ae02-4573-b038-709316bd5f4a)

**Reply by bios4** [reply]
> Is compass accuracy 2/3 understandable or 66% is better?

It mainly depends on the used symbol/icon for the panel I think...
I personally would prefer "2/3".

> How GPS status apps display the compass accuracy value?

I only have one other app in use which displays it - without an icon, but with the whole word accuracy and in meters.
E.g. "Genauigkeit +/- 14 m".

**Reply by devemux86** [reply]
This is about the compass accuracy, that Android reports it in 3 steps.
Not GPS accuracy, which is in meters.

**Reply by devemux86** [reply]
About the icon, we can choose any icon from [here](https://fonts.google.com/icons?icon.set=Material+Icons&icon.style=Filled).

**Reply by bios4** [reply]
> Not GPS accuracy, which is in meters.

I see, I mixed that up...

**Reply by kds112** [reply]
The compass icon in your screenshot looks really suitable for the panel.

**Reply by bios4** [reply]
> About the icon, we can choose any icon from [here](https://fonts.google.com/icons?icon.set=Material+Icons&icon.style=Filled).

Are we limited to the "Material icons (deprecated)" list in this ^^ link, or are also "Material symbols (new)" possible?
There is a `compass_calibration` icon in the deprecated list, which I would definitely NOT find suitable for the panel.

**Reply by kds112** [reply]
I think the Explore icon is suitable and the Explore Off icon could be used if the accuracy is less than 3/3...
![Screenshot_2024-11-11-21-13-57-35_40deb401b9ffe8e1df2f1cc5ba480b12](https://github.com/user-attachments/assets/0663a1db-e3dc-4963-95cc-07964b27b5e3)

**Reply by devemux86** [reply]
Google likes to deprecate things all the time. More in Android. :slightly_smiling_face: 
But the icon collection is still updated and contains many icons with better usage.

The mentioned icon looks like WiFi signal, not much like compass.

**Reply by bios4** [reply]
Does that mean that icons from both lists are possible?
But I agree that `explore` and `explore_off` are quite nice to be used here...

**Comment by kds112** [comment]
Thanks a lot for the implemetation!!
2 minor issues: I start Cruiser, accuracy is 0/3. I calibrate the compass, it changes to 3/3. After a while the accuracy degrades again (checked with a separate compass app), but in Cruiser it stays at 3/3. It should change to 2/3 or even worse.

When I switch off the compass sensor in the settings, the panel stays at 3/3. It should show "-" or something like that.

**Reply by devemux86** [reply]
Switching off the compass setting means it is not used in the app functions.
Or this setting probably is not needed to simplify things.
It does not mean the compass stops existing. Status panels always work.

How the status works, it listens for accuracy reports from the sensor.
If the sensor does not report the changes...
What happens if you restart the app?

**Reply by kds112** [reply]
Restarting the app solved the "problem". I will observe it during my next hiking tour.

I thought switching off the compass sensor would save battery consumtion. If not, the switch is not really necessary...

**Reply by devemux86** [reply]
> Restarting the app solved the "problem".

I don't know if Android allows to ask for the accuracy.
Or it is only through a listener (current implementation), just like GPS.

> If not, the switch is not really necessary.

I have to check this.

**Comment by kds112** [comment]
If Cruiser is in focus the listener works 👍 If Cruiser in the the background and the accuracy degrades the listener seems not to work, because switching Cruiser back in focus the accuracy stays unchanged.

**Reply by devemux86** [reply]
@kds112 Have you excluded Cruiser from all battery / memory optimizations?
See here for instructions: https://dontkillmyapp.com

Maybe the device does not allow the compass to report in background apps.

**Reply by kds112** [reply]
Yes, it is excluded from optimisation. Maybe Cruiser can use the onfocus-event to check the accuracy?

**Reply by devemux86** [reply]
The thing is if there is an Android API to ask for the accuracy on demand
and not via a listener (which apps do not control) like now.

We need to search about it.

**Reply by kds112** [reply]
Maybe this could be helpful for you?
https://github.com/dpa99c/cordova-plugin-compass-accuracy

**Reply by kds112** [reply]
And this one:
https://android-developers.googleblog.com/2024/03/introducing-fused-orientation-provider-api.html?m=1

**Reply by kds112** [reply]
It is a little bit strange. The event listener works now, even in the background in both directions (0/3<->3/3) ... I will test it furthermore... In the moment everything looks okay👍🧭
I changed one setting in the phone in the energy optimisation from "ask always" to "never". Maybe this caused the problem.

**Reply by devemux86** [reply]
> https://android-developers.googleblog.com/2024/03/introducing-fused-orientation-provider-api.html

@kds112 Thanks for the search and the links! An interesting note from this:

_"Fused Orientation Provider updates are only available to apps running in the foreground."_

**Reply by devemux86** [reply]
I thought of an improvement so that it asks for the accuracy again when the app comes back to the foreground.
It will be available in the next update.

**Reply by kds112** [reply]
Perfect! 🎉

**Reply by devemux86** [reply]
> I thought of an improvement so that it asks for the accuracy again when the app comes back to the foreground.

Available in **[Cruiser 5.1.0-b2](https://github.com/devemux86/cruiser/discussions/671)**.

**Reply by kds112** [reply]
At a first glance it works 👍👍 I will test it furthermore at the weekend 🏃

**Comment by kds112** [comment]
I have checked the function on my last hiking tour. It works perfectly 👍🎉 Case closed... 😉🏁

**Reply by devemux86** [reply]
@kds112 Thanks very much for the suggestion and the feedback! :+1:

**Comment by kds112** [comment]
Since the last update b8 (?) the panel does not update properly... I start Cruiser, the accuracy is 0/3. Cruiser is in foreground, I do the "calibration moves". The panel stays at 0/3. Crosscheck with a compass app indicates there 3/3. Going to background, and bringing Cruiser back to foreground does not help. Restarting Cruiser shows 3/3. Can you reproduce the "problem"?

**Reply by devemux86** [reply]
@kds112 Thanks for the report, I will check it.

**Reply by kds112** [reply]
Now it works again... Strange... Maybe Cruiser can check the accuracy level every X seconds? Or do you have a better algorithm? 🤔

**Reply by kds112** [reply]
Or tapping on the panel will call an update routine for the panel?

**Reply by devemux86** [reply]
Apps can only listen to Android's sensors, they can not ask them to refresh themselves.
Android exclusively controls how the sensors work.

Next update will have an improvement and we will see.

**Reply by devemux86** [reply]
This may have been caused after removing the compass sensor setting.
It seems an improved update is needed. :slightly_smiling_face:

**Reply by devemux86** [reply]
@kds112 Please try with **[Cruiser 5.1.0-b9](https://github.com/devemux86/cruiser/discussions/671)**.

**Reply by kds112** [reply]
It looks good 👍 I will check it more intensively on my Sunday hiking tour 🥾🧭

**Reply by kds112** [reply]
Tested on my hiking tour today: It works reliable, thanks a lot!! 👍

**Reply by devemux86** [reply]
@kds112 Thanks for the reports and the feedback! :+1:

**Reply by devemux86** [reply]
@kds112 Is the compass smooth?
Or location rotation if choose the compass orientation?

**Reply by devemux86** [reply]
I cannot reproduce it here, as even with accuracy 2/3 the compass is completely stable.
Even in the very old versions without much filtering.

So you need to test with your devices and report. :slightly_smiling_face:

**Reply by kds112** [reply]
Compass is quite smooth but map rotation is really jittery... I think it was better in the past...



Smartphone is on a wooden table...

**Reply by kds112** [reply]
https://github.com/user-attachments/assets/ff925d1d-b0b2-487e-955d-e44c4476d41d

**Reply by devemux86** [reply]
Videos display compass + map movement together, as it should be, so there is no problem.
Map is rotated based on compass. Compass also moves, but it is less visible due to its size.

If you want the compass (and so the map) to move more rarely, we can increase the filtering.

**Reply by devemux86** [reply]
Try with **[Cruiser 5.1.0-b10](https://github.com/devemux86/cruiser/discussions/671)**.

**Reply by kds112** [reply]
No, please go back to b9... The filtering is not better, but the panel accuracy does not update properly compared to my compass app anymore... You can easily test it moving your smartphone towards a loudspeaker and back...

**Reply by devemux86** [reply]
B9 is also available, you can compare them, we can have one or the other.

@kds112 Does B10 moves the map like in the video?

**Reply by devemux86** [reply]
> panel accuracy does not update properly

@kds112 Panel accuracy code is the same.

**Reply by kds112** [reply]
> B9 is also available, you can compare them, we can have one or the other.
B9 is fine for me. 

> @kds112 Does B10 moves the map like in the video?
Yes, nearly the same.

**Reply by devemux86** [reply]
Well GPS + compass are meant to be used in real conditions and this is outside doing activities.

Not making tests inside on tables. Let's better invest our resources in new important features. :slightly_smiling_face:

**Reply by kds112** [reply]
You are right. The feature with version b9 is really good for hiking. Let's use it👍🥾🧭

**Comment by kds112** [comment]
I am sorry to come back to the panel update. Cruiser b10 always starts with 3/3 even when the accuracy is low (checked with external compass app). Degrading is shown sometimes only, mostly it stays at 3/3. Please, can you check whether the listener does not work properly, especially at app start? Thanks a lot... 👍

**Reply by devemux86** [reply]
On my device the compass accuracy also happens to change over time.

> Cruiser b10 always starts with 3/3

@kds112 Compared to what version?

> even when the accuracy is low

Remember that apps listen to what the sensor reports.
If it shows something, it has come from the sensor.

**Reply by kds112** [reply]
I have checked both b9 and b10. Cruiser always starts with 3/3. In the past b9 had not this effect, now there is the effect, too. Could downgrading from b10 to b9 cause the effect? Maybe it is a problem of my quite old device... I will check it on my wife's phone after the next Release (she does not like beta testing...)

**Reply by devemux86** [reply]
@kds112 Why is this a problem?
It's like asking why GPS reports a particular accuracy.

If the sensor reports something, we should display it.
The default is 0, not 3. The "3" comes from the sensor.

**Reply by kds112** [reply]
The only "problem" is that two different compass apps show low accuracy, but Cruiser shows 3/3 when I start the app without calibrating or moving the phone...

**Reply by devemux86** [reply]
@kds112 I thought of an improvement, which will be available in the next update.

**Reply by kds112** [reply]
In b11 it seems to work properly! 👍 Thanks again a lot!!!

**Comment by devemux86** [comment, marked as answer]
- New panel to display the compass accuracy is available in https://github.com/devemux86/cruiser/discussions/684
